src/snowflake/snowpark/dataframe_transform_functions.py

```diff
@@ -57,6 +69,95 @@ def _validate_formatter_argument(self, fromatter):
         if not callable(fromatter):
             raise TypeError("formatter must be a callable function")
 
+    def _validate_and_extract_time_unit(
+        self, time_str: str, argument_name: str, allow_negative: bool = True
+    ) -> Tuple[int, str]:
+        if not time_str:
+            raise ValueError(f"{argument_name} must not be empty")
+
+        duration = int(time_str[:-1])
+        unit = time_str[-1].lower()
+
+        if not allow_negative and duration < 0:
+            raise ValueError(f"{argument_name} must not be negative.")
+
+        supported_units = ["s", "m", "h", "d", "w", "t", "y"]
+        if unit not in supported_units:
+            raise ValueError(
+                f"Unsupported unit '{unit}'. Supported units are '{supported_units}"
+            )
+
+        # Converting month unit to 'mm' for Snowpark
+        if unit == "t":
+            unit = "mm"
+
+        return duration, unit
+
+    def _get_sliding_interval_start(
+        self, time_col: Column, unit: str, duration: int
+    ) -> Column:
+        unit_seconds = {
+            "s": 1,  # seconds
+            "m": 60,  # minutes
+            "h": 3600,  # hours
+            "d": 86400,  # days
+            "w": 604800,  # weeks
+        }
+
+        if unit == "mm":
+            base_date = lit("1970-01-01").cast("date")
+            months_since_base = months_between(time_col, base_date)
+            current_window_start_month = (months_since_base / duration).cast(
+                "long"
+            ) * duration
+            return to_timestamp(add_months(base_date, current_window_start_month))
+
+        elif unit == "y":
+            base_date = lit("1970-01-01").cast("date")
+            years_since_base = year(time_col) - year(base_date)
+            current_window_start_year = (years_since_base / duration).cast(
+                "long"
+            ) * duration
+            return to_timestamp(add_months(base_date, current_window_start_year * 12))
+
+        elif unit in unit_seconds:
+            # Handle seconds, minutes, hours, days, weeks
+            interval_seconds = unit_seconds[unit] * duration
+            return from_unixtime(
+                (unix_timestamp(time_col) / interval_seconds).cast("long")
+                * interval_seconds
+            )
+
+        else:
+            raise ValueError(
+                "Invalid unit. Supported units are 'S', 'M', 'H', 'D', 'W', 'T', 'Y'."
+            )
+
+    def _perform_aggregations(
+        self,
+        result_df: "snowflake.snowpark.dataframe.DataFrame",
```

Comment:
It's unclear to me why we need `result_df` as an input. Should we rename it to `base_df` maybe?